sched/sched/sched_unlock.c

```diff
@@ -53,60 +53,30 @@
  *
  ****************************************************************************/
 
-#ifdef CONFIG_SMP
-
-int sched_unlock(void)
+void sched_unlock(void)
 {
+  bool need_leave_csection = false;
+  irqstate_t flags;
   FAR struct tcb_s *rtcb;
 
-  /* This operation is safe because the scheduler is locked and no context
-   * switch may occur.
-   */
-
-  rtcb = this_task();
-
-  /* Check for some special cases:  (1) rtcb may be NULL only during
-   * early boot-up phases, and (2) sched_unlock() should have no
-   * effect if called from the interrupt level.
-   */
-
-  if (rtcb != NULL && !up_interrupt_context())
+  if (!up_interrupt_context())
     {
-      /* Prevent context switches throughout the following. */
-
-      irqstate_t flags = enter_critical_section();
```

Comment:
can we hold critical section here once, to avoid the code dup and remove need_leave_csection